### PR TITLE
New `plutarch-extra` functions from Liqwid-Labs

### PR DESCRIPTION
**Pull request by SeungheonOh**
This is first batch of `plutarch-extra` functions from Liqwid-Labs/liqwid-plutarch-extra.

### CODE REVIEW

**Comment by L-as** [comment]
Do you have any tests you could port over?

**Comment by L-as** [comment]
Thanks BTW! Looks great from a first look.

**Comment by SeungheonOh** [comment]
We did have some test, but I don't think they will be compatible with `plutarch-test` because we use `plutarch-quickcheck` for property checking and tasty for unit tests. Not much was there for these functions as well.

**Comment by L-as** [comment]
I don't really think we can merge this without tests. Perhaps we can bring in plutarch-quickcheck, and also clean it up in the process. I was always a bit baffled by your problems with it.

**Comment by SeungheonOh** [comment]
Those functions are copyrighted under Liqwid-Labs. I'll get confirmation from @emiflake to make sure.

Plutarch-Quickcheck 2.0 has been released and it's currently at usable state. The previous problems have only been circumvented by only allowing hask types that can be lifted at this point. However, thanks to the addition of `evalTerm`, we can have efficient shrinkers that doesn't grow almost exponentially in size for non-liftable types (mostly scott encoded lists).

I think with a bit of cleaning we can replace current Hedgehog properties with this one as well. It will be quite quicker than Hedgehog.

**Comment by SeungheonOh** [comment]
I think going with these tests is a good idea for now. Using PQ should be in another PR.

**Review by L-as — COMMENTED** [review]
Haven't finished checking correctness

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 1]
```diff
@@ -21,3 +44,351 @@ pcheckSorted =
...
+  (PIsListLike list a) =>
+  Term
+    s
+    ( (a :--> a :--> PBool)
+        :--> list a
+        :--> list a
+        :--> list a
+    )
+pmergeBy = phoistAcyclic $ pfix #$ plam go
+  where
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/List.hs:79]: `go` is a recursive function. I but I'll make it more clear

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 2]
```diff
@@ -21,3 +44,351 @@ pcheckSorted =
...
+            pif
+              (eq # x # seen)
+              (self # eq # seen # xs)
+              (pcons # seen #$ self # eq # x # xs)
+
+{- | Special version of 'pnubSortBy', which requires elements have 'POrd' instance.
+
+   @since 1.1.0
+-}
+pnubSort ::
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/List.hs:185]: @peter-mlabs or @cjay can you help me on here. I haven't personally seems this used as well.

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 3]
```diff
@@ -21,3 +44,351 @@ pcheckSorted =
...
+
+{- | / O(nlogn) /. Merge sort, bottom-up version, given a custom comparator.
+
+   Assuming the comparator returns true if first value is less than the second one,
+    the list elements will be arranged in ascending order, keeping duplicates in the order
+    they appeared in the input.
+
+   @since 1.1.0
+-}
+pmsortBy ::
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/List.hs:103]: It's generalized for any Plut with `PEq`. I think it's helpful to have

File: plutarch-extra/Plutarch/Extra/List.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/List.hs:79]: `go` is not recursive. The recursion happens at Plutarch-level.

**Comment by SeungheonOh** [comment]
Oh wow `pelimList` is quite a bit faster

File: plutarch-extra/Plutarch/Extra/List.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (peter-mlabs) [reply plutarch-extra/Plutarch/Extra/List.hs:182]: This appears to have been upstreamed by Connor [here](https://github.com/Liqwid-Labs/liqwid-plutarch-extra/commit/7acd1ca7c036c01db778295dcb988d0d3de55275), but `grep` doesn't turn anything up in the current library. I think this is probably safe to toss.

[the remark this answers is not in the document]
Comment (peter-mlabs) [reply plutarch-extra/Plutarch/Extra/List.hs:182]: And looking through the `git log` of Agora, it also looks like this was only ever used in a utils module, and never made its way into the codebase..

[the remark this answers is not in the document]
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/List.hs:182]: I'm dropping this. we can always just compose nub and sort.

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 4]
```diff
@@ -21,3 +44,351 @@ pcheckSorted =
...
+  phoistAcyclic $
+    plam $
+      pelimList (\_ _ -> pcon PTrue) (pcon PFalse)
+
+{- | / O(n) /. Merge two lists which are assumed to be ordered, given a custom comparator.
+    The comparator should return true if first value is less than the second one.
+
+   @since 1.1.0
+-}
+pmergeBy ::
```
[the remark this answers is not in the document]
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/List.hs:68]: `mergyBy f xs ys = zipWith (\x y -> if f x y then x else y) xs ys`, no?

File: plutarch-extra/Plutarch/Extra/List.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/List.hs]: Done

File: plutarch-extra/Plutarch/Extra/List.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/List.hs:82]: We probably want to replace the original with this in a way that doesn't compromise performance by making use of monomorphisation and metaprogramming.

File: plutarch-extra/Plutarch/Extra/List.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/List.hs:51]: no `length (mergeBy f xs ys) == length (xs <> ys)`

[the remark this answers is not in the document]
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/List.hs:51]: Ah yes of course, really, you want pzipWithAndRest or something.

[the remark this answers is not in the document]
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/List.hs:51]: Actually, for efficiency reasons, this is probably best

File: Plutarch/Builtin.hs [hunk 5]
```diff
@@ -518,3 +522,5 @@ instance PTryFrom PData (PAsData PData) where
 instance PTryFrom PData PData where
   type PTryFromExcess PData PData = Const ()
   ptryFrom' opq f = f (opq, ())
+
+instance PTryFrom PData (PAsData PUnit)
```
Comment (L-as) [inline Plutarch/Builtin.hs:526]: What is this instance doing? This seems like nonsense.

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 6]
```diff
@@ -8,16 +28,264 @@ preverse =
...
+
+-- | A special case of 'pmsortBy' which requires elements have 'POrd' instance.
+pmsort ::
+  (POrd a, PIsListLike l a, PIsListLike l (l a)) =>
+  Term s (l a :--> l a)
+pmsort = phoistAcyclic $ pmsortBy # comp
+  where
+    comp = phoistAcyclic $ plam (#<)
+
+-- | /O(n log n)/. Remove dupicate elements in a list with the given function to decide duplicate.
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:128]: dupicate

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 7]
```diff
@@ -8,16 +28,264 @@ preverse =
...
+    s
+    ( (a :--> a :--> PBool)
+        :--> list a
+        :--> list a
+    )
+pnubBy = phoistAcyclic $
+  plam $ \eq l ->
+    pelimList (\x xs -> go # eq # x # xs) l l
+  where
+    go = phoistAcyclic pfix #$ plam $ \self eq seen l ->
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:142]: you're hoisting `pfix`...

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 8]
```diff
@@ -8,16 +28,264 @@ preverse =
...
+        ( \x xs ->
+            pif
+              (eq # x # seen)
+              (self # eq # seen # xs)
+              (pcons # seen #$ self # eq # x # xs)
+        )
+        (psingleton # seen)
+        l
+
+-- | Special version of 'pnub', which requires elements have 'POrd' instance.
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:153]: `POrd`?

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 9]
```diff
@@ -8,16 +28,264 @@ preverse =
...
+  plam $ \eq l ->
+    pelimList (\x xs -> go # eq # x # xs) l l
+  where
+    go = phoistAcyclic pfix #$ plam $ \self eq seen l ->
+      pelimList
+        ( \x xs ->
+            pif
+              (eq # x # seen)
+              (self # eq # seen # xs)
+              (pcons # seen #$ self # eq # x # xs)
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:148]: You're fixing at the wrong place. You want to fix after applying the equality function.

File: plutarch-extra/Plutarch/Extra/List.hs (hunk 6 above)
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:128]: This assumes the input is sorted, and it's also `O(n)` AFAICT.

Comment (L-as) [reply plutarch-extra/Plutarch/Extra/List.hs:128]: If you want it to assume sortedness, express that invariant in the types somehow.
A newtype seems sensible, albeit I remember seeing a paper for a new approach on this. Can't remember the name.

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 10]
```diff
@@ -8,16 +28,264 @@ preverse =
...
+  Term
+    s
+    ( (a :--> a :--> PBool)
+        :--> list a
+        :--> PBool
+    )
+pisUniqBy = phoistAcyclic $
+  plam $ \eq xs ->
+    let nubbed = pnubBy # eq # xs
+     in plength # xs #== plength # nubbed
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:175]: Very inefficient. This also assumes sortedness I suppose?

File: Plutarch/Builtin.hs [hunk 11]
```diff
@@ -273,6 +273,10 @@ instance PIsData a => PlutusType (PAsData a) where
   pcon' (PAsData t) = punsafeCoerce $ pdata t
   pmatch' t f = f (PAsData $ pfromData $ punsafeCoerce t)
 
+instance (PIsData a) => PIsData (PAsData a) where
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply Plutarch/Builtin.hs:276]: @chfunghr wasn't this for some API types right?

File: plutarch-extra/Plutarch/Extra/List.hs [hunk 12]
```diff
@@ -8,16 +28,264 @@ preverse =
...
+pmsort ::
+  (POrd a, PIsListLike l a, PIsListLike l (l a)) =>
+  Term s (l a :--> l a)
+pmsort = phoistAcyclic $ pmsortBy # comp
+  where
+    comp = phoistAcyclic $ plam (#<)
+
+-- | /O(n log n)/. Remove dupicate elements in a list with the given function to decide duplicate.
+pnubBy ::
+  forall (a :: S -> Type) (s :: S) list.
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/List.hs:130]: PType anytime we have this.

**Comment by SeungheonOh** [comment]
Liqwid-Labs internally, we found out some(a lot) of functions here are quite poorly designed and we have decided to rewrite/update them all together. Therefore, this PR will be closed, and there will be new PRs with revised functionalities. 

Sorry for inconvenience.

**Pull request closed without merging**